text/0069-api-v2.md

```diff
@@ -0,0 +1,171 @@
+
+### Client Version
+
+TiKV clients will have a new configuration choosing `API V1` or `API V2` mode on connecting to TiKV.
+
+TiKV will reject the request if the client mode is mismatched to the TiKV configuration. There is one exception: With TiKV in `API V2` mode, you can use `TxnClient` in `API V1` mode to access the TiDB data (key starting with `m` or `t`) which is invisible to the client in `API V2` mode.
+
+Note that in `API V2` mode, you can't specify CF in `RawClient`.
+
+Using TxnKV along with RawKV is safe in `API V2` mode.
+
+Listed below is the detailed compatibility matrix:
+
+|              | V1 Server | V2 Server |
+| ------------ | --------- | --------- |
+| V1 RawClient | Raw Data  | Forbidden |
+| V1 TxnClient | Txn Data  | TiDB Data |
+| V2 RawClient | Forbidden | Raw Data  |
+| V2 TxnClient | Forbidden | Txn Data  |
+
+### Keyspace
+
+When API V2 is enabled for the client, the user will be able to specify a keyspace for a session of `RawClient` or `TxnCient`. A keyspace is identified by a string, whose default value is 'default'.
+
+Keys in other keyspace will be invisible to the user.
+
+## Reference-level design
+
+### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
+
+The `{keyspace prefix id}` is the [keyspace](https://github.com/tikv/rfcs/pull/39) prefix for seperating keys of different keyspace. It should be an vary-length integer whose highest bit of every byte denotes whether the next byte is still part of the integer. The client will fetch the prefix from PD by the keyspace name specified by the user when initializing the client, so it means that the keyspace prefix is valid during the session, in other words, the change on keyspace on PD will not take effect on running seesions.
+
+Note that in TxnKV, the key will be encoded by `Memory Comparable Encoding`. But since the `Memory Comparable Encoding` will not change the starting bytes and only add paddings, there won't be any overlap between RawKV and TxnKV.
+
+### RawKV Value Encode
+
+If the key is started with `r` or `a`, then the value can be either:
+
+1. `{0x0}{data}`
+2. `{0x1}{TTL expire timestamp}{data}`
+
+### PD
+
+Add a new http interface for adding, renaming, deleteing and querying the mapping from keyspace name to prefix id:
+
+```javascript
+// list all keyspaces
+GET /keyspaces
+[
+  {
+    name: "default",
+    id: "0"
+  },
+  {
+    name: "redis",
+    id: "1"
+  }
+]
+
+// add new keyspace
+POST /keyspaces
+{
+  name: "mongodb",
+  // optional, pd will allocate one if not specified
+  id: "2"
```

Comment:
should it be auto-generated by tikv, which can avoid possible coniflict?

Reply:
The name and id are both required to be unique. Allowing specifying id so that user can recover a delete keyspace.

Reply:
> recover a delete keyspace

when will we remove the data after deleting a keyspace?

Reply:
No one but the user can delete the data. So the data will be left there if the user delete the keyspace without clearing them.

Reply:
> No one but the user can delete the data. So the data will be left there if the user deletes the keyspace without clearing them.

@andylokandy after the keyspace is deleted, the key-values inside that key-space can not be able to be accessed anymore, Am I correct?

Could you elaborate on why not deleting the key-values inside that keyspace?

Reply:
I also agree that current design is the most appropriate at time. Not reclaiming data when key space is deleted sounds a reasonable protection to user in case there is mistake during operation. If we really would like to improve this, we can add retain policy just like what K8S does for PVC and POD.